hudi-common/src/main/java/org/apache/hudi/common/table/log/block/HoodieLogBlockFactory.java

```diff
@@ -0,0 +1,90 @@
+import org.apache.hudi.common.table.HoodieTableConfig;
+import org.apache.hudi.exception.HoodieException;
+import org.apache.hudi.metadata.HoodieMetadataHFileDataBlock;
+import org.apache.hudi.metadata.HoodieMetadataPayload;
+import org.apache.hudi.metadata.HoodieTableMetadata;
+
+import java.util.List;
+import java.util.Map;
+
+public class HoodieLogBlockFactory {
+
+  /**
+   * Util method to get a data block for the requested type.
+   *
+   * @param logDataBlockFormat - Data block type
+   * @param recordList         - List of records that goes in the data block
+   * @param header             - data block header
+   * @param tableConfig        - Table config
+   * @param metadataConfig     - Metadata config
+   * @param tableBasePath      - Table base path
+   * @param populateMetaFields - Whether to populate meta fields in the record
+   * @return Data block of the requested type.
+   */
+  public static HoodieLogBlock getBlock(HoodieLogBlock.HoodieLogBlockType logDataBlockFormat,
+                                        List<IndexedRecord> recordList,
+                                        Map<HoodieLogBlock.HeaderMetadataType, String> header,
+                                        HoodieTableConfig tableConfig, HoodieMetadataConfig metadataConfig,
+                                        String tableBasePath, boolean populateMetaFields) {
+    final boolean isMetadataKeyDeDuplicate = metadataConfig.getRecordKeyDeDuplicate()
+        && HoodieTableMetadata.isMetadataTable(tableBasePath);
+    String keyField;
+    if (populateMetaFields) {
+      keyField = (isMetadataKeyDeDuplicate
+          ? HoodieMetadataPayload.SCHEMA_FIELD_ID_KEY : HoodieRecord.RECORD_KEY_METADATA_FIELD);
+    } else {
+      keyField = tableConfig.getRecordKeyFieldProp();
+    }
+    return getBlock(logDataBlockFormat, recordList, header, keyField, isMetadataKeyDeDuplicate);
+  }
+
+  /**
+   * Util method to get a data block for the requested type.
+   *
+   * @param logDataBlockFormat       - Data block type
+   * @param recordList               - List of records that goes in the data block
+   * @param header                   - data block header
+   * @param keyField                 - FieldId to get the key from the records
+   * @param isMetadataKeyDeDuplicate - Whether metadata key de duplication needed
+   * @return Data block of the requested type.
+   */
+  private static HoodieLogBlock getBlock(HoodieLogBlock.HoodieLogBlockType logDataBlockFormat, List<IndexedRecord> recordList,
+                                         Map<HoodieLogBlock.HeaderMetadataType, String> header, String keyField,
+                                         boolean isMetadataKeyDeDuplicate) {
+    switch (logDataBlockFormat) {
+      case AVRO_DATA_BLOCK:
+        return new HoodieAvroDataBlock(recordList, header, keyField);
+      case HFILE_DATA_BLOCK:
+        if (isMetadataKeyDeDuplicate) {
+          return new HoodieMetadataHFileDataBlock(recordList, header, keyField);
```

Comment:
lets log leak metadata concepts here?

Reply:
sure, I can pull the key deduplication logic to higher layer. Will post a revision.

Reply:
@vinothchandar Refactored the code based on our last discussion. Please take one more look.

Also, added the unit test in a chained PR https://github.com/manojpec/hudi/pull/4